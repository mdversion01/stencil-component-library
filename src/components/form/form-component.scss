form {
  margin: 0;
  padding: 0;
  width: 100%;
}

legend {
  padding: 0 10px;
}

legend.left {
  text-align: left;
}
legend.right {
  text-align: right;
}
legend.center {
  text-align: center;
}

fieldset {
  border: 1px solid #ccc;
  border-radius: 5px;
}
