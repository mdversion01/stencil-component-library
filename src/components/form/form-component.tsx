// src/components/form/form-component.tsx
import { Component, h, Prop } from '@stencil/core';

@Component({
  tag: 'form-component',
  styleUrls: ['../layout-styles.scss', '../form-styles.scss'],
  shadow: false,
})
export class FormComponent {
  /** Native form attributes */
  @Prop() action: string = '';
  @Prop() method: string = '';

  /** Render a fieldset wrapper (with optional legend) */
  @Prop() fieldset: boolean = false;
  @Prop() legend: boolean = false;

  // ⚠ Use inline union types so components.d.ts doesn’t reference unknown names
  @Prop() legendPosition: 'left' | 'center' | 'right' | string = 'left';
  @Prop() legendTxt: string = 'Add Title Here';

  /** Layout + identity used by slotted children */
  @Prop() formLayout: '' | 'horizontal' | 'inline' = '';
  @Prop() formId: string = '';

  /**
   * If true, do not render a <form>. Inputs placed outside the form can still
   * read formId via closest('form-component') and set their own form attribute.
   */
  @Prop() outsideOfForm: boolean = false;

  /** Optional border/box styling */
  @Prop() bcolor: string = '';
  @Prop() bradius?: number;
  @Prop() bstyle: string = '';
  @Prop() bwidth?: number;

  /** Additional inline styles to append (CSS string) */
  @Prop() styles: string = '';

  private computedCssText(): string {
    const parts: string[] = [];
    if (this.bcolor) parts.push(`border-color:${this.bcolor};`);
    if (this.bstyle) parts.push(`border-style:${this.bstyle};`);
    if (this.bwidth != null && !Number.isNaN(this.bwidth)) parts.push(`border-width:${this.bwidth}px;`);
    if (this.bradius != null && !Number.isNaN(this.bradius)) parts.push(`border-radius:${this.bradius}px;`);
    const dynamic = parts.join(' ');
    const extra = (this.styles || '').trim();
    return [extra, dynamic].filter(Boolean).join(' ');
  }

  private layoutClass(): string | undefined {
    if (this.formLayout === 'horizontal') return 'horizontal';
    if (this.formLayout === 'inline') return 'inline';
    return undefined;
  }

  private renderSlot() {
    // Children use closest('form-component') to read formId/formLayout.
    return <slot name="formField"></slot>;
  }

  private renderBase() {
    if (this.outsideOfForm) {
      // No invalid <div form="...">; optionally expose a data attribute if you want a hook.
      return <div data-form-id={this.formId || undefined}>{this.renderSlot()}</div>;
    }
    return (
      <form
        class={this.layoutClass()}
        action={this.action || undefined}
        id={this.formId || undefined}
        method={this.method || undefined}
      >
        {this.renderSlot()}
      </form>
    );
    }

  private renderWithFieldset(cssText: string) {
    if (this.outsideOfForm) {
      return (
        <fieldset style={{ cssText }}>
          {this.legend ? <legend class={this.legendPosition || 'left'}>{this.legendTxt || 'Add Title Here'}</legend> : null}
          {this.renderSlot()}
        </fieldset>
      );
    }
    return (
      <form
        class={this.layoutClass()}
        action={this.action || undefined}
        id={this.formId || undefined}
        method={this.method || undefined}
      >
        <fieldset style={{ cssText }}>
          {this.legend ? <legend class={this.legendPosition || 'left'}>{this.legendTxt || 'Add Title Here'}</legend> : null}
          {this.renderSlot()}
        </fieldset>
      </form>
    );
  }

  render() {
    const cssText = this.computedCssText();
    return this.fieldset ? this.renderWithFieldset(cssText) : this.renderBase();
  }
}
