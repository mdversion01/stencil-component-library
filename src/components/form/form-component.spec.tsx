// src/components/form/form-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { FormComponent } from './form-component';

describe('<form-component>', () => {
  it('renders a <form> with basic attributes', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component action="/submit" method="post" form-id="contact"></form-component>`,
    });

    const root = page.root!;
    const form = root.querySelector('form') as HTMLFormElement;

    expect(form).toBeTruthy();
    expect(form.getAttribute('action')).toBe('/submit');
    expect(form.getAttribute('method')).toBe('post');
    expect(form.getAttribute('id')).toBe('contact');
    // no layout class by default
    expect(form.className).toBe('');
  });

  it('applies layout classes (horizontal / inline)', async () => {
    const pageH = await newSpecPage({
      components: [FormComponent],
      html: `<form-component form-layout="horizontal"></form-component>`,
    });
    expect(pageH.root!.querySelector('form')!.classList.contains('horizontal')).toBe(true);

    const pageI = await newSpecPage({
      components: [FormComponent],
      html: `<form-component form-layout="inline"></form-component>`,
    });
    expect(pageI.root!.querySelector('form')!.classList.contains('inline')).toBe(true);
  });

  it('renders outsideOfForm container instead of <form>', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component outside-of-form form-id="external"></form-component>`,
    });

    const form = page.root!.querySelector('form');
    const div = page.root!.querySelector('div');

    expect(form).toBeNull();
    expect(div).toBeTruthy();
    // exposes data-form-id for children to read if needed
    expect(div!.getAttribute('data-form-id')).toBe('external');
  });

  it('renders fieldset + legend with defaults', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component fieldset legend></form-component>`,
    });

    const fieldset = page.root!.querySelector('fieldset') as HTMLFieldSetElement;
    const legend = fieldset.querySelector('legend') as HTMLLegendElement;

    expect(fieldset).toBeTruthy();
    expect(legend).toBeTruthy();
    expect(legend.textContent).toBe('Add Title Here');
    // default legend position class
    expect(legend.className).toBe('left');
  });

  it('renders fieldset inside <form> when not outsideOfForm, with custom legend & styles', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component
             fieldset
             legend
             legend-txt="Profile"
             legend-position="center"
             bcolor="#333"
             bstyle="solid"
             bwidth="2"
             bradius="8"
             styles="margin: 4px;"
           ></form-component>`,
    });

    const form = page.root!.querySelector('form')!;
    const fieldset = form.querySelector('fieldset') as HTMLFieldSetElement;
    const legend = fieldset.querySelector('legend') as HTMLLegendElement;

    expect(form).toBeTruthy();
    expect(fieldset).toBeTruthy();
    expect(legend.textContent).toBe('Profile');
    expect(legend.className).toBe('center');

    // cssText normalization varies (spaces after ':' etc.). Use regex-friendly checks.
    const css = (fieldset.style && fieldset.style.cssText) || fieldset.getAttribute('style') || '';

    // keep a simple contains check for the custom styles string
    expect(css.replace(/\s+/g, ' ')).toContain('margin: 4px;');

    // allow optional whitespace after ':' and before ';'
    expect(css).toMatch(/border-color:\s*#333\s*;/);
    expect(css).toMatch(/border-style:\s*solid\s*;/);
    expect(css).toMatch(/border-width:\s*2px\s*;/);
    expect(css).toMatch(/border-radius:\s*8px\s*;/);
  });

  it('renders slotted children in the "formField" slot', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component>
               <div slot="formField" id="child-a">A</div>
               <div slot="formField" id="child-b">B</div>
             </form-component>`,
    });

    const formOrDiv = page.root!.querySelector('form, div')!;
    const a = formOrDiv.querySelector('#child-a');
    const b = formOrDiv.querySelector('#child-b');

    expect(a).toBeTruthy();
    expect(b).toBeTruthy();
    expect(a!.textContent).toBe('A');
    expect(b!.textContent).toBe('B');
  });

  // ---------------- SNAPSHOTS ----------------

  it('matches snapshot (base form)', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component action="/s" method="get" form-id="f1"></form-component>`,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('matches snapshot (fieldset + legend + styles, outsideOfForm)', async () => {
    const page = await newSpecPage({
      components: [FormComponent],
      html: `<form-component
               fieldset
               legend
               legend-position="right"
               legend-txt="Settings"
               outside-of-form
               form-id="external"
               bcolor="red"
               bstyle="dashed"
               bwidth="1"
               bradius="4"
               styles="padding:2px;"
             >
               <div slot="formField">Inner field</div>
             </form-component>`,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });
});
