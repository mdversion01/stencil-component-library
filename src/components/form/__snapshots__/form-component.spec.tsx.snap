// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<form-component> matches snapshot (base form) 1`] = `
<form-component action="/s" form-id="f1" method="get">
  <!---->
  <form action="/s" id="f1" method="get"></form>
</form-component>
`;

exports[`<form-component> matches snapshot (fieldset + legend + styles, outsideOfForm) 1`] = `
<form-component bcolor="red" bradius="4" bstyle="dashed" bwidth="1" fieldset="" form-id="external" legend="" legend-position="right" legend-txt="Settings" outside-of-form="" styles="padding:2px;">
  <!---->
  <fieldset style="padding: 2px; border-color: red; border-style: dashed; border-width: 1px; border-radius: 4px;">
    <legend class="right">
      Settings
    </legend>
    <div slot="formField">
      Inner field
    </div>
  </fieldset>
</form-component>
`;
